### PR TITLE
Avoid allocating when performing VisitRulesFor on service accounts

### PR DESCRIPTION
**Pull request by smarterclayton**
Service account authorization checks are done frequently and were
observed to perform 7% of allocations on a system running e2e tests.
The allocation comes from when we walk the authorization rules to
find matching service accounts.

Optimize the check for service account names to avoid allocating.

Allocations over about 5m of a 1.12 server running e2es.

image

```
4892626	5.21%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/authentication/serviceaccount.MakeUsername	(inline)
0	0.00%	5.21%	4869687	5.19%	net/http.HandlerFunc.ServeHTTP	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac.(*RBACAuthorizer).Authorize	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/registry/rbac/validation.appliesToUser	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/registry/rbac/validation.appliesTo	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/registry/rbac/validation.(*DefaultRuleResolver).VisitRulesFor	
0	0.00%	5.21%	4869687	5.19%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1	
0	0.00%	5.21%	4869687	5.19%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/server/filters.WithCORS.func1	
0	0.00%	5.21%	4861494	5.18%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1	
0	0.00%	5.21%	4869687	5.19%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1	
0	0.00%	5.21%	4869687	5.19%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1	
0	0.00%	5.21%	4869687	5.19%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1	
0	0.00%	5.21%	4869687	5.19%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAudit.func1	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/authorization/union.unionAuthzHandler.Authorize	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/vendor/k8s.io/apiserver/pkg/authorization/union.(*unionAuthzHandler).Authorize	
0	0.00%	5.21%	4892626	5.21%	github.com/openshift/origin/pkg/authorization/authorizer/browsersafe.(*browserSafeAuthorizer).Authorize	
```

/kind bug

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go [hunk 1]
```diff
@@ -36,6 +36,27 @@ func MakeUsername(namespace, name string) string {
 	return ServiceAccountUsernamePrefix + namespace + ServiceAccountUsernameSeparator + name
 }
 
+// MatchesUsername checks whether the provided username matches the namespace and name without
+// allocating. Use this when checking a service account namespace and name against a known string.
+func MatchesUsername(namespace, name string, username string) bool {
+	if !strings.HasPrefix(username, ServiceAccountUsernamePrefix) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: can avoid calling HasPrefix() twice (once directly, once indirectly via TrimPrefix) by taking the substring yourself once you pass HasPrefix checks.

File: staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go [hunk 2]
```diff
@@ -36,6 +36,27 @@ func MakeUsername(namespace, name string) string {
...
+		return false
+	}
+	username = strings.TrimPrefix(username, ServiceAccountUsernamePrefix)
+
+	if !strings.HasSuffix(username, name) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: alternative prefix/suffix is hard to follow... consider checking prefix and consuming ServiceAccountUsernamePrefix, then namespace, then ServiceAccountUsernameSeparator, then name?

**Comment by yliaog** [comment]
Looks good, what are the resulting improvement with this optimization? how much allocations saved?

File: staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go (hunk 1 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: It's not super expensive in this case because most of the prefix is in the cache.  The cost is the allocation rather than the scan.

File: staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: Algorithmic improvement to check prefix then suffix.

**Comment by smarterclayton** [comment]
All of the allocations.

File: staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go (hunk 1 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: The less ugly way to check it would be to compare to the previous length, but that makes the code harder to read.

File: staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: It might be a wash across namespaces, but in the common ones (all in the same namespace) it's a lot.  I won't go to the wall for it.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: if it makes a difference, I can live with it

Comment (smarterclayton) [reply staging/src/k8s.io/apiserver/pkg/authentication/serviceaccount/util.go]: Actually, I did the simpler thing which bypasses TrimPrefix and is probably faster than both because we're doing a linear scan.
